### ISSUE TITLE
[Refactor] Refactor Flow to use save & restore pattern for handling flow switching

### ISSUE THREAD

**Issue by MaxGraey** [AssemblyScript/assemblyscript#2506]
See original proposal in this [comment](https://github.com/AssemblyScript/assemblyscript/pull/2352#discussion_r963046874)
But instead `lock` / `unlock` perhaps better to use `save` / `restore` names

**Comment by HerrCai0907** [comment]
But for some complex cases such as `doCompileWhileStatement`, there are lots of flow. How can we know which flow should be `restore` and restore what?

**Comment by MaxGraey** [comment]
Good point. Need to think more. Perhaps `save` can accept optional argument. Or we can use stack-based approach for save / restore

**Comment by dcodeIO** [comment]
> Or we can use stack-based approach

Indeed, some sort of `push`/`pop` for flows seems fitting. Other than that, I am a little skeptical because it restricts what we can do, so we should be reasonably sure that more complicated flow ordering won't be necessary.

**Comment by HerrCai0907** [comment]
> I am a little skeptical because it restricts what we can do, so we should be reasonably sure that more complicated flow ordering won't be necessary.
I think tree may be suitable for it. Fork some children from parent flow and then go back to the parent

**Comment by MaxGraey** [comment]
Flow is pretty linear stuff. It can go into deep inner scope, but I don't think we need tree or DAG for handling this
